### PR TITLE
Onestop id lineage

### LINKED ISSUE #332: Onestop ID lineage/deaccessioning
changing the Onestop ID on an entity
- [x] a new changeset action: `changeOnestopID`
- [x] when the changeset is applied, it sets `action=change_onestop_id` in the `old_` entities record
- [x] `find_by_onestop_id` methods will now search through both the `current_` and `old_` entities records. Will redirect to the entity with the current Onestop ID

destroying an entity with a Onestop ID
- uses the existing `destroy` changeset action
- [x] when the changeset is applied, it sets `action=destroy` in the `old_` entities record
- [x] `find_by_onestop_id` methods will now search through both the `current_` and `old_` entities records before returning their results. If a Onestop ID has been deleted, the API will indicate that.

merging entities with multiple Onestop IDs into one
- [x] add new `merge`changeset action that takes in an array of Onestop IDs and one entity/Onestop ID for the target/survivor
- [x] when the changeset is applied, set `action=merge` in the `old_` entity records and point their `current` column to the target/survivor
- [x] `find_by_onestop_id` methods will now search through both the `current_` and `old_` entities records before returning their results. For merged records, redirect to the target/survivor

- [ ] add callbacks to merge and change. Specifically, when Feed onestop id changes, the S3 link might need updating (related to #895; may not be an issue), and the stop_pattern onestop id in the RouteStopPattern stop_pattern attributes needs to be updated. 

After this is in place:
- When geometry changes outside of a Onestop ID's geohash, automatically change the geohash

### PR DESCRIPTION
**Pull request by doublestranded**
closes #332

### CODE REVIEW

File: app/models/concerns/current_tracked_by_changeset.rb [hunk 1]
```diff
@@ -50,13 +55,64 @@ def apply_changes_create_update(changeset: nil, changes: nil, cache: {})
...
+
+    def apply_changes_change_onestop_id(changeset: nil, changes: nil, cache: {})
+      changes.each do |change|
+        unless change.has_key?(:new_onestop_id)
+          raise Changeset.Error.new(changeset, "could not find newOnestopId")
+        end
+        existing_model = find_existing_model(change.merge({ onestop_id: change[:onestop_id] }))
+        if existing_model
+          attrs_to_apply = apply_params(existing_model.as_change.merge({ onestop_id: change[:new_onestop_id] }), cache)
+          new_model = self.create_making_history(changeset: changeset, new_attrs: attrs_to_apply)
```
Comment (irees) [inline app/models/concerns/current_tracked_by_changeset.rb]: existing_model.update_making_history(changeset: changeset, new_attrs: {onestop_id: change[:new_onestop_id]})

File: app/models/concerns/current_tracked_by_changeset.rb [hunk 2]
```diff
@@ -130,13 +178,15 @@ def before_destroy_making_history(changeset, old_model)
     return true
   end
 
-  def destroy_making_history(changeset: nil)
+  def destroy_making_history(changeset: nil, action: nil, current: nil)
```
Comment (irees) [inline app/models/concerns/current_tracked_by_changeset.rb]: can we use old_attrs here like update_making_history

File: app/models/concerns/current_tracked_by_changeset.rb [hunk 3]
```diff
@@ -38,10 +45,51 @@ def apply_change_create_update(changeset: nil, change: nil, cache: {})
...
+
+    def apply_change_merge_onestop_ids(changeset: nil, change: nil, onestop_ids_to_merge: nil, cache: {})
+      attrs_to_apply = apply_params(change, cache)
+      merge_into_model = find_by_onestop_id(change[:onestop_id])
+      if merge_into_model
+        merge_into_model.update_making_history(changeset: changeset, new_attrs: attrs_to_apply)
+      else
+        merge_into_model = self.create_making_history(changeset: changeset, new_attrs: attrs_to_apply)
+        merge_into_model.after_merge_onestop_ids(onestop_ids_to_merge, changeset)
+        merge_into_model.update_associations(changeset)
```
Comment (irees) [inline app/models/concerns/current_tracked_by_changeset.rb]: ian: check if update_associations is needed here

File: app/models/concerns/has_a_onestop_id.rb [hunk 4]
```diff
@@ -5,28 +5,47 @@ module HasAOnestopId
     validates :onestop_id, presence: true, uniqueness: true
     validate :onestop_id, :validate_onestop_id
 
+    def self.current_from_old_reference(old_entity, find_method)
+      unless ['merge', 'change_onestop_id'].exclude?(old_entity.action) || old_entity.current.nil?
+        self.send(find_method, { onestop_id: old_entity.current.onestop_id })
+      end
+    end
+
     def self.find_by_onestop_id!(onestop_id)
```
Comment (doublestranded) [inline app/models/concerns/has_a_onestop_id.rb]: Split into 6 methods total:
Keeping the 2 previously implemented find_by_onestop_id methods
2 new methods named "find_by_current_and_old_onestop_id" (add comments to explain)
2 "plural" find_by_onestop_ids methods that maintain the existing behavior

File: spec/models/changeset_spec.rb [hunk 5]
```diff
@@ -205,17 +205,171 @@
...
+              onestopId: Stop.first.onestop_id
+            }
+          }
+        ]
+      })
+      changeset.apply!
+      expect(Stop.find_by_onestop_id!(merge_stop_1.onestop_id)).to eq Stop.first
+      expect(Stop.find_by_onestop_id!(merge_stop_2.onestop_id)).to eq Stop.first
+      expect(Stop.all).to match_array([Stop.find_by_onestop_id!('s-9q8yt4b-1AvHoS')])
+      expect(OldStop.find(2)).to eq OldStop.find_by_onestop_id!(merge_stop_1.onestop_id)
```
Comment (doublestranded) [inline spec/models/changeset_spec.rb]: Don't use hardcoded id

File: spec/models/changeset_spec.rb [hunk 6]
```diff
@@ -205,17 +205,171 @@
...
+      expect(OldStop.last).to eq OldStop.find_by_onestop_id!(merge_stop_2.onestop_id)
+      expect(OldStop.last.current).to eq Stop.first
+      expect(OldStop.last.action).to eq 'merge'
+    end
+
+    it 'updates rsp stop pattern onestop ids on merge onestop ids' do
```
Comment (doublestranded) [inline spec/models/changeset_spec.rb]: stop onestop ids

**Comment by irees** [comment]
:+1:

**Pull request merged**